спринт 14

SleeplessShu opened the pull request:
- подключен плеер
- добавлен отсчёт времени воспроизведения
- автоматизирована активация поиска
- добавлен прогрессбар

SleeplessShu committed:
- подключен плеер
- добавлен отсчёт времени воспроизведения
- автоматизирована активация поиска
- добавлен прогрессбар

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/ActivityPlayer.kt:
```diff
@@ -1,43 +1,58 @@
+        private const val STATE_PREPARED = 1
+        private const val STATE_PLAYING = 2
+        private const val STATE_PAUSED = 3
     }
```
Согласно [код-стайлу](https://kotlinlang.org/docs/coding-conventions.html#class-layout) Котлина Companion Object должен находится в конце класса.
Хотя я лично знаю команды, где он пишется в начале класса.
В целом, код-стайл это не про жёсткие правила, а про договорённости внутри
команды. Но я бы рекомендовал использовать стандартные практики в новых
проектах, где ещё не устоялись нестадартные. Это будет очень полезно при
онбординге новых разработчиков.

В целом рекомендую внимательно ознакомиться с Coding Conventions Котлина и по
возможности им следовать.

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/ActivityPlayer.kt:
```diff
@@ -1,43 +1,58 @@
         val radiusPx = formatDpToPx(this, 2)
-        //временное переключение кнопок
+
+        //---->временное переключение кнопок<----
```
В целом в разработке считается хорошей практикой писать комментарии на 
английском. Посколько неизвестно кто будет читать твой код, коллеги могут быть 
и не русскоговорящими.
Но если с английским не очень хорошо, то не стоит волноваться. Это абсолютно 
необязательно, в больших компаниях я видел комментарии как на русском так и на 
английском.

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/ActivityPlayer.kt:
```diff
@@ -1,43 +1,58 @@
+    private var playingTimeCounter: Runnable? = null
+    private var playerState = STATE_DEFAULT
+    private lateinit var mediaPlayer: MediaPlayer
+    private var trackTime by Delegates.notNull<Long>()
```
Честно говоря, первый раз вижу использование этого делегата. Выглядит просто как fancy `lateinit`. По сути, настолько же опасная вещь, так же легко отсрелить себе ногу.

SleeplessShu replied:
использовал решение, предложенное средой разработки. его смысл не понимаю, заменил на инициализацию "0"

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/ActivityPlayer.kt:
```diff
@@ -1,43 +1,58 @@
+
+        mainThreadHandler = Handler(Looper.getMainLooper())
+
         val currentTrack: TrackData? = intent.getParcelableExtra(GET_TRACK_DATA_FROM_SEARCH)
```
Этот метод уже `deprecated`. Но это довольно известная штука и обычно все
все пользуются уловкой, про которую можно прочитать на [StackOverflow](https://stackoverflow.com/questions/73019160/the-getparcelableextra-method-is-deprecated)

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/ActivitySearch.kt:
```diff
@@ -56,8 +62,10 @@ class ActivitySearch : ActivityBase() {
             TrackAdapter(sharedPreferencesManager, emptyList(), TrackAdapter.VIEW_TYPE_EMPTY) {
-                trackAdapter.updateTracks(emptyList(), TrackAdapter.VIEW_TYPE_EMPTY)
-                searchTracks(binding.tiSearch.text.toString())
+                if (clickDebounce()) {
```
Не очень ясно чему именно служит этот debounce.

SleeplessShu committed:
Привет, Руслан! Спасибо за подробное изучение моего проекта и крутые правки.

внесены изменения:

- добавлено скругление картинки
- lateinit delagate заменено на инициализацию нулём
- getParcelableExtra заменен на serializable
- добавлен класс  DebounceClickListener
- к кнопкам во всех активити добавлены дебоунсеры втч по элементу списка
- круглые кнопки в плеере реагируют круглыми foreground
- исправлена ошибка состояний плеера генерящая тосты
- таймер при окончании трека сбрасывается на нули
- кнопка “очистить историю” исчезает при начале поиска

я предполагаю, что сценарии должны быть такими:

- При нажатии на крестик очищается строка поиска, последняя выдача заменяется историей.
- При пустом поисковом запросе отображается история.
- При нажатии на кнопку поиска, поиск по дебоунсу не сработает

kekulta commented on app/src/main/java/com/practicum/playlistmaker_ver2/ActivitySearch.kt:
```diff
@@ -106,22 +103,18 @@ class ActivitySearch : ActivityBase() {
         binding.bEraseText.setOnClickListener {
             binding.queryInput.text.clear()
-            trackAdapter.updateTracks(emptyList(), TrackAdapter.VIEW_TYPE_EMPTY)
+            //trackAdapter.updateTracks(emptyList(), TrackAdapter.VIEW_TYPE_EMPTY)
```
Закоментированный код мы не коммитим!

kekulta commented:
Забыл отдельно отметить иконку! Очень здорово, что добавил. Добавляет в проект чувство завершённости.
